include/AssembleElemSolverAlgorithm.h

```diff
@@ -40,6 +54,60 @@ class AssembleElemSolverAlgorithm : public SolverAlgorithm
   virtual void initialize_connectivity();
   virtual void execute();
 
+  template<typename LambdaFunction>
+  void run_algorithm(stk::mesh::BulkData& bulk_data, LambdaFunction lambdaFunc)
+  {
+    stk::mesh::MetaData& meta_data = bulk_data.mesh_meta_data();
+    const int lhsSize = rhsSize_*rhsSize_;
+    const int scratchIdsSize = rhsSize_;
+
+   const int bytes_per_team = 0;
+   const int bytes_per_thread = calculate_shared_mem_bytes_per_thread(lhsSize, rhsSize_, scratchIdsSize,
+                                                                    meta_data.spatial_dimension(), dataNeededByKernels_);
+   stk::mesh::Selector s_locally_owned_union =
+           meta_data.locally_owned_part() & stk::mesh::selectUnion(partVec_);
+ 
+   stk::mesh::BucketVector const& elem_buckets =
+           realm_.get_buckets(entityRank_, s_locally_owned_union );
+ 
+   auto team_exec = sierra::nalu::get_team_policy(elem_buckets.size(), bytes_per_team, bytes_per_thread);
+   Kokkos::parallel_for(team_exec, [&](const sierra::nalu::TeamHandleType& team)
+   {
+     stk::mesh::Bucket & b = *elem_buckets[team.league_rank()];
+ 
+     ThrowAssertMsg(b.topology().num_nodes() == (unsigned)nodesPerEntity_,
+                    "AssembleElemSolverAlgorithm expected nodesPerEntity_ = "
+                    <<nodesPerEntity_<<", but b.topology().num_nodes() = "<<b.topology().num_nodes());
+ 
+     SharedMemData smdata(team, bulk_data, dataNeededByKernels_, nodesPerEntity_, rhsSize_);
+ 
+     const stk::mesh::Entity* entityNodes[simdLen];
+     const size_t bucketLen   = b.size();
+     const size_t simdBucketLen = get_simd_bucket_length(bucketLen);
+ 
+     Kokkos::parallel_for(Kokkos::TeamThreadRange(team, simdBucketLen), [&](const size_t& bktIndex)
+     {
+       int numSimdElems = get_next_num_elems_simd(bktIndex, bucketLen);
+       smdata.numSimdElems = numSimdElems;
+ 
+       for(int simdElemIndex=0; simdElemIndex<numSimdElems; ++simdElemIndex) {
+         stk::mesh::Entity element = b[bktIndex*simdLen + simdElemIndex];
+         entityNodes[simdElemIndex] = bulk_data.begin_nodes(element);
+         fill_pre_req_data(dataNeededByKernels_, bulk_data, element,
+                           *smdata.prereqData[simdElemIndex], interleaveMEViews_);
+       }
+ 
+       copy_and_interleave(smdata.prereqData, numSimdElems, smdata.simdPrereqData, interleaveMEViews_);
+ 
+       if (!interleaveMEViews_) {
+         fill_master_element_views(dataNeededByKernels_, bulk_data, smdata.simdPrereqData);
+       }
+
+       lambdaFunc(smdata, nodesPerEntity_, entityNodes);
```

Comment:
Let's say that we have a element-based scheme (all hex-mesh) activated with an 1) interior contribution, a 2) wall function, and a 3) symmetry/open. 

Is the concept/design point that we would have three activated ElemAlgs with each being 1) as an interior, 2) a face bc, and 3) two-kernels of a face/element?

If we add a hybrid mesh - well - let's not go there yet:)

Reply:
I'm not sure. I was just reading through AssembleMomentumElemOpenSolverAlg again, trying to think about these kind of questions. I think the 'run_algorithm' method could be the basis for an elem alg or a face alg, with little or no changes. But perhaps a face/elem alg needs a separate 'run_face_elem_algorithm' which would loop over face buckets, get the elem for each face, and manage faceScratchViews and elemScratchViews.